### PR TITLE
SDL automator scripts

### PR DESCRIPTION
**Pull request by jonfortescue**
~First draft of scripts -- don't merge yet as I haven't added TSA upload functionality yet. I want to get early feedback tho and see if anyone hates the design I'm going with.~

TSA functionality implemented. Should be good to go, though likely there will be snags and bugs that will need to be worked out, so we should definitely test them more thoroughly from release definitions.

The design here is that these scripts would be stuck inside a nuget package and referenced from release definitions.

Example usage: 

```powershell
.\execute-sdl.ps1 -GuardianCliLocation $env:GuardianInstallLocation -Repository "dotnet/arcade" -SourceDirectory $env:SourceDirectory -ArtifactsDirectory $env:ArtifactsDirectory -DncEngAccessToken $env:DncEngAccessToken -SourceToolsList @("credscan") -ArtifactToolsList @("binskim")
```

### CODE REVIEW

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Param(
```
Comment (tmat) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: I'd avoid abbreviating `Guardian` as `gdn` (I assume that's what it stands for).
Also use `DncEngAccessToken` instead of `DncengPat` for clarity.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  [string] $DncengPat,
+  [string] $PushReason
+)
+
+$sdlDir = "$SourcesDirectory/../sdl-tool-cfg"
+if (Test-Path $sdlDir) {
+  Remove-Item -Force -Recurse $sdlDir
+}
+
+Write-Host "git clone https://dnceng:$DncengPat@dev.azure.com/dnceng/internal/_git/sdl-tool-cfg $sdlDir"
```
Comment (tmat) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: I don't think we should be printing out access token, right?

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 [hunk 3]
```diff
@@ -0,0 +1,26 @@
+Param(
+  [string] $Repository,
+  [string] $SourcesDirectory,
+  [string] $DncengPat,
+  [string] $GdnLoggerLevel="Standard"
+)
+
```
Comment (alexperovich) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: Every powershell script should have this at the top
```powershell
$ErrorActionPreference = "Stop"
Set-StrictMode -Version 2.0
```

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+$zipFile = "$SourcesDirectory/gdn.zip"
+
+Add-Type -AssemblyName System.IO.Compression.FileSystem
+Try
+{
+  Write-Host "Downloading gdn folder from internal config repostiory..."
+  Invoke-WebRequest -Headers @{ "Accept"="application/zip"; "Authorization"="Basic $encodedPat" } -Uri $uri -OutFile $zipFile
+  [System.IO.Compression.ZipFile]::ExtractToDirectory($zipFile, $sourcesDirectory)
+} Catch [System.Net.WebException] {
+  # if the folder does not exist, we'll do a gdn init and create a PR for it
```
Comment (alexperovich) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: This isn't creating a PR, don't say it is.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 (hunk 1 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: Not abbreviating guardian strictly speaking -- this is referring to pushing the .gdn folder. Will do for the pat name.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 (hunk 2 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: Yeah, definitely good practice even if the releases are always internal.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  Write-Host "Initializing Guardian..."
+  Write-Host "guardian init --working-directory $SourcesDirectory --logger-level $GdnLoggerLevel"
+  guardian init --working-directory $SourcesDirectory --logger-level $GdnLoggerLevel
```
Comment (alexperovich) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: You need to check exit codes everywhere you call an executable from powershell.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 6]
```diff
@@ -0,0 +1,23 @@
+Param(
+  [string] $Repository,
+  [string] $SourcesDirectory,
+  [string] $GdnFolder,
+  [string] $DncengPat,
+  [string] $PushReason
+)
+
+$sdlDir = "$SourcesDirectory/../sdl-tool-cfg"
```
Comment (alexperovich) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: Why are you leaving the source directory? This dir won't be cleaned by azure pipelines.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 3 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: will do

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 4 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: oops, left over from previous version. Thanks for the catch.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 5 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: Ah you're correct. will do.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+if (Test-Path $sdlDir) {
+  Remove-Item -Force -Recurse $sdlDir
+}
+
+Write-Host "git clone https://dnceng:$DncengPat@dev.azure.com/dnceng/internal/_git/sdl-tool-cfg $sdlDir"
+git clone https://dnceng:$DncengPat@dev.azure.com/dnceng/internal/_git/sdl-tool-cfg $sdlDir
+Copy-Item -Recurse -Force $GdnFolder $SourcesDirectory/../sdl-tool-cfg/$Repository
```
Comment (alexperovich) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]:
```suggestion
Copy-Item -Recurse -Force $GdnFolder $sdlDir/$Repository
```

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 (hunk 6 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: I will do something different.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  [string] $DncEngAccessToken,
+  [string] $GuardianLoggerLevel="Standard"
+)
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version 2.0
+
+# Construct basic auth from AzDO access token; construct URI to the repository's gdn folder stored in that repository; construct location of zip file
+$encodedPat = [Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes(":$DncEngAccessToken"))
+$uri = "https://dev.azure.com/dnceng/internal/_apis/git/repositories/sdl-tool-cfg/Items?path=$([Uri]::EscapeDataString("/$Repository/.gdn"))&versionDescriptor[versionOptions]=0&`$format=zip&api-version=5.0-preview.1"
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: >$([Uri]::EscapeDataString("/$Repository/.gdn")) [](start = 93, length = 47)

Maybe want to pull this double parenthetical quoted thing out of the middle of this string into a variable, to make the URI a bit easier on the eyes.

Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: >versionDescriptor[versionOptions]=0 [](start = 141, length = 35)

How is this going to handle versioning/branching? We seem to just be pulling... I assume master?  We aren't really specificying any version at all... what is it going to download?

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  if (Test-Path $gdnFolder) {
+    # Remove the gdn folder if it exists (it shouldn't unless there's too much caching; this is just in case)
+    Remove-Item -Force -Recurse $gdnFolder
+  }
+  [System.IO.Compression.ZipFile]::ExtractToDirectory($zipFile, $WorkingDirectory)
+} Catch [System.Net.WebException] {
+  # if the folder does not exist, we'll do a guardian init and push it to the remote repository
+  Write-Host "Initializing Guardian..."
+  Write-Host "$GuardianCliLocation init --working-directory $WorkingDirectory --logger-level $GuardianLoggerLevel"
+  Invoke-Expression "$GuardianCliLocation init --working-directory $WorkingDirectory --logger-level $GuardianLoggerLevel"
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: >$GuardianCliLocation [](start = 21, length = 20)

Invoke-Expression is... messy.  Turning things into string only to then reparse them as command should be avoided when possible... Why not 
```
&$GuardianCliLocation init --working-directory $WorkingDirectory --logger-level $GuardianLoggerLevel
```

That has the added benefit of not exploding if "WorkingDirectory" has a space in it. :-)

File: src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1 [hunk 10]
```diff
@@ -0,0 +1,37 @@
+Param(
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1]: "run-sdl" and "execute-sdl" are pretty similar names... What is the difference between the two, and could we name one of them more clearly?

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+# We copy the .gdn folder from our local run into the git repository so it can be committed
+$sdlRepositoryFolder = Join-Path (Join-Path $sdlDir $Repository) ".gdn"
+if (Get-Command Robocopy) {
+  Robocopy /S $GdnFolder $sdlRepositoryFolder
+} else {
+  rsync -r $GdnFolder $sdlRepositoryFolder
+}
+# cd to the sdl-config directory so we can run git there
+$currentDirectory = (Get-Location).Path
+Set-Location $sdlDir
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: >Set-Location $sdlD [](start = 0, length = 18)

Why not pushd/popd instead of managing the location stack yourself?

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 [hunk 12]
```diff
@@ -0,0 +1,50 @@
+Param(
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: Is branching an issue for this script? I assume yes? Or do all branches have the same config/baselines?

**Comment by ChadNedzlek** [comment]
I'm not 100% sure where the "PAT" is supposed to come from? Is this only going to run on dnceng/internal repos, and it will just be part of the build definitions for those repositories?

**Comment by jonfortescue** [comment]
@ChadNedzlek yeah, the PAT will be provided as part of the build defs since it will only run in internal. That's part of how we'll prevent it from running in the public instance.

thanks for the feedback btw! I'll go through and fix this after I'm done with FR.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 8 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: So right now there's no direct plan for handling versioning/branching but I imagine it wouldn't necessarily be difficult. Just add folders for each branch basically. I think I will modify this for now to put the code in a folder called "master" so that that functionality can be added in the future more easily.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 9 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: Will do. Thanks for the feedback!

File: src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1 (hunk 10 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1]: they do need to be clarified. execute is the top-level that runs all the other scripts rather than making you call them individually.

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 (hunk 11 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: didn't realize I could use those in powershell!

File: src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1 (hunk 12 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/push-gdn.ps1]: All branches have the same configs/baselines for now. Per your comment above I'm going to stick this stuff one subfolder deeper to make it easier for implementing versioning/branching in the future if necessary.

File: src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1 (hunk 8 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/init-sdl.ps1]: Nvm added this anyway.

File: src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1 (hunk 10 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/run-sdl.ps1]: Resolved by rename.

File: eng/common/sdl/run-sdl.ps1 [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version 2.0
+$LASTEXITCODE = 0
+
+# We store config files in the r directory of .gdn
+Write-Host $ToolsList
+$gdnConfigPath = Join-Path $GdnFolder "r"
+$gdnConfig = ""
```
Comment (sunandabalu) [inline eng/common/sdl/run-sdl.ps1]: This variable is not used anywhere, why is this needed?

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+} else {
+  $guardianCliLocation = $GuardianCliLocation
+}
+
+& $(Join-Path $PSScriptRoot "init-sdl.ps1") -GuardianCliLocation $guardianCliLocation -Repository $Repository -BranchName $BranchName -WorkingDirectory $ArtifactsDirectory -DncEngAccessToken $DncEngAccessToken -GuardianLoggerLevel $GuardianLoggerLevel
+$gdnFolder = Join-Path $ArtifactsDirectory ".gdn"
+
+if ($TsaOnboard) {
+  if ($TsaCodebaseName -and $TsaNotificationEmail -and $TsaCodebaseAdmin -and $TsaBugAreaPath) {
+    Write-Host "$guardianCliLocation tsa-onboard --codebase-name `"$TsaCodebaseName`" --notification-alias `"$TsaNotificationEmail`" --codebase-admin `"$TsaCodebaseAdmin`" --instance-url `"https://dev.azure.com/dnceng/`" --project-name `"internal`" --area-path `"$TsaBugAreaPath`" --working-directory $ArtifactsDirectory --logger-level $GuardianLoggerLevel"
```
Comment (sunandabalu) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: The instance-url and project-name should not be hard-coded as different teams might want to have different locations to file bugs, this should be a param as well.

File: eng/common/sdl/run-sdl.ps1 (hunk 13 above)
Comment (jonfortescue) [reply eng/common/sdl/run-sdl.ps1]: Looks like an accidental holdover from a previous version of the script. Good catch.

File: eng/common/sdl/run-sdl.ps1 [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+$gdnConfigPath = Join-Path $GdnFolder "r"
+
+foreach ($tool in $ToolsList) {
+  $gdnConfigFile = Join-Path $gdnConfigPath "$tool-configure.gdnconfig"
+  $config = $False
+  Write-Host $tool
+  # We have to manually configure tools that run on source to look at the source directory only
+  if ($tool -eq "credscan") {
+    Write-Host "$GuardianCliLocation configure --working-directory $WorkingDirectory --tool $tool --output-path $gdnConfigFile --logger-level $GuardianLoggerLevel --noninteractive --force --args `" TargetDirectory : $TargetDirectory `""
+    &$GuardianCliLocation configure --working-directory $WorkingDirectory --tool $tool --output-path $gdnConfigFile --logger-level $GuardianLoggerLevel --noninteractive --force --args " TargetDirectory : $TargetDirectory "
```
Comment (ChadNedzlek) [inline eng/common/sdl/run-sdl.ps1]: Is it possible for whatever "GaurdianCliLocation" is pointing to to fail?  We seem to mostly just plow along assuming that it did what we asked without error.

Comment (jonfortescue) [reply eng/common/sdl/run-sdl.ps1]: @ChadNedzlek -- you're right, we should add some checks in there. Will do that.

Comment (sunandabalu) [reply eng/common/sdl/run-sdl.ps1]: Added check to validate GuardianCLILocation

**Comment by sunandabalu** [comment]
@tmat, @alexperovich , @ChadNedzlek We have addressed all comments, is there any more feedback ? If not, we can merge this change.

File: eng/common/sdl/run-sdl.ps1 (hunk 15 above)
Comment (ChadNedzlek) [reply eng/common/sdl/run-sdl.ps1]: I don't see any evidence this has happened.

Comment (sunandabalu) [reply eng/common/sdl/run-sdl.ps1]: The check is in the execute-sdl.ps1 which calls the run-sdl.ps1. I'll explicitly add here as well.

Comment (ChadNedzlek) [reply eng/common/sdl/run-sdl.ps1]: The check I'm worried about is if this command fails, we assume it's true and continue running the script. There is no if ($LASTEXITCODE) here to make sure the command we ran actually executed successfully.

Comment (sunandabalu) [reply eng/common/sdl/run-sdl.ps1]: Added $LASTEXITCODE check to configure and run commands

File: eng/common/sdl/run-sdl.ps1 [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version 2.0
+$LASTEXITCODE = 0
+
+# We store config files in the r directory of .gdn
+Write-Host $ToolsList
+$gdnConfigPath = Join-Path $GdnFolder "r"
+
+if (Test-Path $GuardianCliLocation) {
```
Comment (ChadNedzlek) [inline eng/common/sdl/run-sdl.ps1]: I'd invert this check + early return (though "Write-Error", because of the ErrorActionPreference, is functionally an early return).

Then we wouldn't have to have the if enclose the whole file, meaning I would get to like 56 and then wonder, "Wait, what is this else and else to?"

Comment (sunandabalu) [reply eng/common/sdl/run-sdl.ps1]: Makes sense, I'll fix that.

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  [string] $TsaBugAreaPath,                             # Optional: only needed if TsaOnboard is true; the area path where TSA will file bugs in AzDO
+  [string] $TsaIterationPath,                           # Optional: only needed if TsaOnboard is true; the iteration path where TSA will file bugs in AzDO
+  [string] $GuardianLoggerLevel="Standard"              # Optional: the logger level for the Guardian CLI; options are Trace, Verbose, Standard, Warning, and Error
+)
+
+$ErrorActionPreference = "Stop"
+Set-StrictMode -Version 2.0
+$LASTEXITCODE = 0
+
+$RepoName = $Repository -replace '(.*?)-(.*)', '$1/$2';
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: This assumes repositories are named `x-y`, but that's not true for all repos. Why do we need to transform this name? From the usage seems like using the original name should be ok.

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  [string] $GuardianCliLocation,                        # Optional: Direct location of Guardian CLI executable if GuardianPackageName & NugetPackageDirectory are not specified
+  [string] $Repository,                                 # Required: the name of the repository (e.g. dotnet/arcade)
+  [string] $BranchName="master",                        # Optional: name of branch or version of gdn settings; defaults to master
+  [string] $SourceDirectory,                            # Required: the directory where source files are located
+  [string] $ArtifactsDirectory,                         # Required: the directory where build artifacts are located
+  [string] $DncEngAccessToken,                          # Required: access token for dnceng; should be provided via KeyVault
+  [string[]] $SourceToolsList,                          # Optional: list of SDL tools to run on source code
+  [string[]] $ArtifactToolsList,                        # Optional: list of SDL tools to run on built artifacts
+  [bool] $TsaPublish=$False,                            # Optional: true will publish results to TSA; only set to true after onboarding to TSA
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: >TSA [](start = 97, length = 3)

IT's not clear what TSA means

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+  $guardianCliLocation = $GuardianCliLocation
+}
+
+$ValidPath = Test-Path $guardianCliLocation
+
+if ($ValidPath -eq $False)
+{
+  Write-Error "Invalid Guardian CLI Location."
+}
+else{
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: call `exit 1` instead of having `else { .. }` clause

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+& $(Join-Path $PSScriptRoot "init-sdl.ps1") -GuardianCliLocation $guardianCliLocation -Repository $RepoName -BranchName $BranchName -WorkingDirectory $ArtifactsDirectory -DncEngAccessToken $DncEngAccessToken -GuardianLoggerLevel $GuardianLoggerLevel
+$gdnFolder = Join-Path $ArtifactsDirectory ".gdn"
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: What is this directory used for?

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+if ($TsaOnboard) {
+  if ($TsaCodebaseName -and $TsaNotificationEmail -and $TsaCodebaseAdmin -and $TsaBugAreaPath) {
+    Write-Host "$guardianCliLocation tsa-onboard --codebase-name `"$TsaCodebaseName`" --notification-alias `"$TsaNotificationEmail`" --codebase-admin `"$TsaCodebaseAdmin`" --instance-url `"$TsaInstanceUrl`" --project-name `"$TsaProjectName`" --area-path `"$TsaBugAreaPath`" --iteration-path `"$TsaIterationPath`" --working-directory $ArtifactsDirectory --logger-level $GuardianLoggerLevel"
+    & $guardianCliLocation tsa-onboard --codebase-name "$TsaCodebaseName" --notification-alias "$TsaNotificationEmail" --codebase-admin "$TsaCodebaseAdmin" --instance-url "$TsaInstanceUrl" --project-name "$TsaProjectName" --area-path "$TsaBugAreaPath" --iteration-path "$TsaIterationPath" --working-directory $ArtifactsDirectory --logger-level $GuardianLoggerLevel
+    if ($LASTEXITCODE -ne 0) {
+      Write-Error "Guardian tsa-onboard failed with exit code $LASTEXITCODE."
+    }
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: >} [](start = 4, length = 1)

`exit $LASTEXITCODE` ?

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+  } else {
+    Write-Error "Could not onboard to TSA -- not all required values ($$TsaCodebaseName, $$TsaNotificationEmail, $$TsaCodebaseAdmin, $$TsaBugAreaPath) were specified."
+  }
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: >} [](start = 2, length = 1)

`exit 1`?

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+      $TsaRepositoryName = "$($Repository)-$($BranchName)"
+    }
+    Write-Host "$guardianCliLocation tsa-publish --all-tools --repository-name `"$TsaRepositoryName`" --branch-name `"$TsaBranchName`" --build-number `"$BuildNumber`" --codebase-name `"$TsaCodebaseName`" --notification-alias `"$TsaNotificationEmail`" --codebase-admin `"$TsaCodebaseAdmin`" --instance-url `"$TsaInstanceUrl`" --project-name `"$TsaProjectName`" --area-path `"$TsaBugAreaPath`" --iteration-path `"$TsaIterationPath`" --working-directory $SourceDirectory --logger-level $GuardianLoggerLevel"
+    & $guardianCliLocation tsa-publish --all-tools --repository-name "$TsaRepositoryName" --branch-name "$TsaBranchName" --build-number "$BuildNumber" --codebase-name "$TsaCodebaseName" --notification-alias "$TsaNotificationEmail" --codebase-admin "$TsaCodebaseAdmin" --instance-url "$TsaInstanceUrl" --project-name "$TsaProjectName" --area-path "$TsaBugAreaPath" --iteration-path "$TsaIterationPath" --working-directory $ArtifactsDirectory  --logger-level $GuardianLoggerLevel
+    if ($LASTEXITCODE -ne 0) {
+      Write-Error "Guardian tsa-publish failed with exit code $LASTEXITCODE."
+    }
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: >} [](start = 4, length = 1)

`exit $LASTEXITCODE`

File: eng/common/sdl/execute-all-sdl-tools.ps1 [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+  } else {
+    Write-Error "Could not publish to TSA -- not all required values ($$TsaBranchName, $$BuildNumber) were specified."
+  }
```
Comment (tmat) [inline eng/common/sdl/execute-all-sdl-tools.ps1]: >} [](start = 2, length = 1)

`exit 1`?

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 17 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: The repo name in release definition comes from the env variable Build.Repository.Name which is intentionally stored with a '-' for eg: dotnet-arcade, dotnet-arcade-services.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 18 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: TSA is an automation framework that Guardian uses to file bugs in VSTS.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 20 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: .gdn folder is where all the tool configs, run results are stored. This is generated initially when Guardian is initiated

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 18 above)
Comment (tmat) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: It'd be good to state that in the comment.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 17 above)
Comment (tmat) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: Yes, those are the mirrorred repos. Sure, but why does it need to be translated to `dotnet/arcade`?
Is this not gonna be available for non-mirrorred repos?

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 20 above)
Comment (tmat) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: Perhaps it could be called `GuardianResults` then?

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 18 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: Will do

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 20 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: Unfortunately this is generated by the guardian CLI and all its commands looks for the folder with that name.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 17 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: We are using the repo name and the branch name to persist the gdn folder for baselining purposes, the repo name being in the dotnet/arcade format. I'll fix the scripts to be able to accept both formats. But yes, currently the scripts are only for internal mirrored repo.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 19 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: I have changed the Write-Error to Write-Host and exit 1 coz just exit is not saying why it exited.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 21 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: I have changed the Write-Error to Write-Host and exit $LASTEXITCODE coz just exit is not saying why it exited.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 22 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: I have changed the Write-Error to Write-Host and exit 1 coz just exit is not saying why it exited.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 23 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: I have changed the Write-Error to Write-Host and exit $LASTEXITCODE coz just exit is not saying why it exited.

File: eng/common/sdl/execute-all-sdl-tools.ps1 (hunk 24 above)
Comment (sunandabalu) [reply eng/common/sdl/execute-all-sdl-tools.ps1]: I have changed the Write-Error to Write-Host and exit 1 coz just exit is not saying why it exited.
